File: Loom.Providers/OpenAI/OpenAIToolSchemaBuilder.vb
Imports Loom.Core.Models
Imports Loom.Providers.Core

Namespace OpenAI
    ''' <summary>
    ''' Reconstructs raw data into the specific OpenAI format
    ''' </summary>
    Public Class OpenAIToolSchemaBuilder

        Public Function Build(def As ToolDefinition) As ProviderToolSchema
            ' Check parameters and required != Nothing
            Dim props As New Dictionary(Of String, ProviderPropertySchema)()

            If def.Parameters IsNot Nothing AndAlso def.Parameters.Count > 0 Then
                For Each p In def.Parameters
                    Dim propSchema As New ProviderPropertySchema()
                    propSchema.type = If(String.IsNullOrEmpty(p.Type), "string", p.Type)
                    propSchema.description = If(p.Description, "")

                    ' Enum added only if != Nothing
                    If p.Enums IsNot Nothing AndAlso p.Enums.Count > 0 Then
                        propSchema.[enum] = p.Enums
                    End If

                    props.Add(p.Name, propSchema)
                Next
            End If

            Dim requiredList As List(Of String) = If(def.Required IsNot Nothing, def.Required, New List(Of String)())

            Dim functionSchema As New ProviderFunctionSchema With {
                .type = "object",
                .properties = props,
                .required = requiredList,
                .additionalProperties = False
            }

            Return New ProviderToolSchema With {
                .name = def.ToolName,
                .description = def.Description,
                .parameters = functionSchema,
                .strict = def.Strictness.GetValueOrDefault(False)
            }
        End Function



    End Class

End Namespace

File: implementations/csharp/src/Loom.Providers.OpenAI/OpenAIToolSchemaBuilder.cs
using Loom.Core.Models;
using Loom.Providers.OpenAI.Schemas;
using System.Collections.Generic;

namespace Loom.Providers.OpenAI
{
    /// <summary>
    /// Reconstructs raw data into the specific OpenAI format
    /// </summary>
    public class OpenAIToolSchemaBuilder
    {
        public OpenAIToolSchema Build(ToolDefinition def)
        {
            // Check parameters and required != null
            var props = new Dictionary<string, OpenAIPropertySchema>();

            if (def.Parameters != null && def.Parameters.Count > 0)
            {
                foreach (var param in def.Parameters)
                {
                    var propSchema = new OpenAIPropertySchema();
                    propSchema.type = string.IsNullOrEmpty(param.Type) ? "string" : param.Type;
                    propSchema.description = param.Description ?? "";

                    // Enum added only if != null
                    if (param.Enums != null && param.Enums.Count > 0)
                        propSchema.@enum = param.Enums;

                    props.Add(param.Name, propSchema);
                }
            }

            var requiredList = def.Required ?? new List<string>();

            var functionSchema = new OpenAIFunctionSchema
            {
                type = "object",
                properties = props,
                required = requiredList,
                additionalProperties = false
            };

            return new OpenAIToolSchema
            {
                name = def.ToolName,
                description = def.Description,
                parameters = functionSchema,
                strict = def.Strictness.GetValueOrDefault(false)
            };
        }
    }
}
